mirror_maker.go

```diff
@@ -214,15 +203,13 @@ func (this *MirrorMaker) startProducers() {
 	}
 }
 
-func (this *MirrorMaker) produceRoutine(producer *sarama.Producer, channelIndex int) {
+func (this *MirrorMaker) produceRoutine(producer Producer, channelIndex int) {
 	for msg := range this.messageChannels[channelIndex] {
-		var key sarama.Encoder
-		if !this.config.PreservePartitions {
-			key = sarama.ByteEncoder(msg.Key)
+		if this.config.PreservePartitions {
+			producer.Input() <- &ProducerMessage{Topic: this.config.TopicPrefix + msg.Topic, Key: uint32(msg.Partition), Value: msg.Value, KeyEncoder: &Int32Encoder{}, ValueEncoder: this.config.ValueEncoder}
```

Comment:
Suppose it would be better to move the `&Int32Encoder{}` out of the range statement and reuse this instance each time rather than creating a new one every time